xlators/mgmt/glusterd/src/glusterd-utils.c

```diff
@@ -5472,6 +5472,8 @@ glusterd_compare_friend_data(dict_t *peer_data, dict_t *cmp, int32_t *status,
         dict_unref(arg->peer_ver_data);
         GF_FREE(arg);
     }
+    if (!update && arg)
+        GF_FREE(arg);
```

Comment:
The way this function is implemented makes this check more complex. The loop at lines 5444..5458 may jump to `out` even after having set `update` to true, so in this case we still need to destroy `arg`.

Additionally, arg contains some data, like `peer_data` and `peer_ver_data` that needs to be unref'd before destroying `arg`itself.

You can also merge the right condition with the code at lines 5470..5474 to make it more compact.

Unless you know very well how this function works (which I don't) and why the loop doesn't terminate when `update` is set to true, probably the easiest and safest way would be to set `arg = NULL` after line 5466, and here we only need to check if `arg != NULL` before destroying it.

Reply:
My understanding is as follows:
The while loop runs for all the available peers in the cluster and compares the volume data in both the peers. And, it only exit's if get RJT or `glusterd_compare_friend_volume` returns an error.
Also, the arg isn't destroyed in case we set `update` as `true`, as the function, `glusterd_launch_synctask` is an async function and goes on to execute even after we have finished the execution of the `glusterd_compare_friend_data` function (the same is highlighted in the comment at line: 5461)

I think the fix could be to only change the condition at line: 5470 to `if ((ret && arg) || (GLUSTERD_VOL_COMP_RJT == *status))`

Reply:
@mohit84 can you also please share your views on the working of this section of the code and the proposed fix?

Reply:
I think we need to remove GLUSTERD_VOL_COMP_RJT check from a loop. We should continue the loop to try 
    for next volume. As per current code is if  current node volume cksum is not matching with peer cksum it rejects all the other volumes updates also, this behavior is not correct.We should consider volumes updates for those cksum are matching so i believe there is no requirement to check the condition (GLUSTERD_VOL_COMP_RJT == *status)).
After remove this condition there is no issue to cleanup arg memory.

Reply:
Isn't this check happening in case of a peer probe event?
If that's the case, it should be failing in case of a conflict rather than getting connected with `Peer rejected`, which is what I get as per the code in function `glusterd_ac_handle_friend_add_req` which is calling this function at line: 988 in glusterd_sm.c file.

Reply:
@mohit84 if you could please share your view on the above comment, we can make the changes accordingly in this PR.

Reply:
The correct fix is we need to cleanup memory for arg only while update is not true otherwise the memory will be clean by the function glusterd_import_friend_volumes_synctask so you can change the condition to cleanup memory for arg 
at the line 5470 only if (!update && arg) instead of (ret && arg)

Reply:
Incase of failure, we should be freeing it as well, isn't it?
Then, that would make the check as `((ret || !update) && arg)`

Reply:
Yes we can add a check for ret value also.